New conversion script

comex opened the pull request:
Current status: Working, except for the latest GPTQ-for-LLaMa format that includes `g_idx`.  This turns out to require changes to GGML, so for now it only works if you use the `--outtype` option to dequantize it back to f16 (which is pointless except for debugging).

I also included some cleanup for the C++ code.

This script is meant to replace all the existing conversion scripts (including the ones that convert from older GGML formats), while also adding support for some new formats.  Specifically, I've tested with:

- [x] `LLaMA` (original)
- [x] `llama-65b-4bit`
- [x] `alpaca-native`
- [x] `alpaca-native-4bit`
- [x] LLaMA converted to 'transformers' format using `convert_llama_weights_to_hf.py`
- [x] `alpaca-native` quantized with `--true-sequential --act-order --groupsize 128` (dequantized only)
- [x] same as above plus `--save_safetensors`
- [x] GPT4All
- [x] stock unversioned ggml
- [x] ggmh
- [x] alpaca-30b-4bit.pt
- [x] alpaca-30b-4bit.safetensors
- [x] alpaca-30b-4bit-128g.safetensors
- [x] koala-13B-HF
- [x] koala-13B-4bit-128g.safetensors (dequantized only)
- [ ] koala-13B-4bit-128g.pt

There's enough overlap in the logic needed to handle these different cases that it seemed best to move to a single script.

I haven't tried this with Alpaca-LoRA because I don't know where to find it.

Useful features:

- Uses multiple threads for a speedup in some cases (though the Python GIL limits the gain, and sometimes it's disk-bound anyway).

- Combines split models into a single file (both the intra-tensor split of the original and the inter-tensor split of 'transformers' format files).  Single files are more convenient to work with and more friendly to future changes to use memory mapping on the C++ side.  To accomplish this without increasing memory requirements, it has some custom loading code which avoids loading whole input files into memory at once.

- Because of the custom loading code, it no longer depends in PyTorch, which might make installing dependencies slightly easier or faster... although it still depends on NumPy and sentencepiece, so I don't know if there's any meaningful difference.  In any case, I also added a requirements.txt file to lock the dependency versions in case of any future breaking changes.

- Type annotations checked with mypy.

- Some attempts to be extra user-friendly:

    - The script tries to be forgiving with arguments, e.g. you can specify either the model file itself or the directory containing it.

    - The script doesn't depend on config.json / params.json, just in case the user downloaded files individually and doesn't have those handy.  But you still need tokenizer.model and, for Alpaca, added_tokens.json.

    - The script tries to give a helpful error message if added_tokens.json is missing.

anzz1 commented:
Looks fantastic! :tada: 

Agree that the conversion scripts should be merged as one. 

Have you checked that the sha256 checksums match for files produced with the old and new scripts? :thinking: 

Minor comment, I think naming it like `convert-model-to-ggml.py` would be more verbose, as the name `convert.py` doesn't really tell its' purpose.

green-s commented:
Could you maybe add [safetensors](https://github.com/huggingface/safetensors) support? People are starting to distribute GPTQ weights in that format instead, since it doesn't allow arbitrary code execution. Usually it's just a matter of using `safetensors.torch.load_from_file` in place of `torch.load` but since you're not using torch.load it might be a bit trickier.

Belluxx commented:
I converted with the script a  7B 3.77GB 4bit gptq (no grops) model. The converted file however is 5.39GB. Is this expected?

It's also very slow compared to the RTN q4 model because it swaps on the disk now due to its size.

BadisG commented:
You should verify if your script works with the new techniques proposed by @qwopqwop200 
https://github.com/qwopqwop200/GPTQ-for-LLaMa
![image](https://user-images.githubusercontent.com/110173477/228200224-16ec21d1-5f7d-490e-8816-0cead3b51332.png)

I think it's not the case as someone reported an error here : https://github.com/ggerganov/llama.cpp/issues/442#issuecomment-1486553066

luxtiasco commented:
The parameters "true-sequential" and "act-order" have been proposed for some time now, but they have only become fully functional in the recent commits.

BadisG commented:
@luxtiasco It's not finished yet, once qwopqwop200 will be able to make "act-order" and "groupsize 128" work together, we'll get a really great quantization 😄👍

BadisG commented:
https://github.com/qwopqwop200/GPTQ-for-LLaMa/commit/4e141a8c4725958b187f9d5d1ffac13f22a204c9

The madman did it! Now it's possible to get both groupsize and act-order !

ggerganov commented:
🦙 !

I'm fully OK with this change - cannot comment on the Python code as I don't have experience
cc @jart - this change allows to generate single-file models from the get-go. Might have some relevance for the `mmap` stuff, so bringing your attention just in case

jart commented:
Looks very promising! Single file models would be nice. The main thing I want is for the tensors to be mmap()'able. In order for that to happen, multi-dimensional tensors need to be laid out in the file in such a way that they don't need to be reshaped in order to be loaded. The memory layout on disk, should be the same as what ggml wants in memory at runtime. The format should also observe ideally a 32-byte alignment. Does this change do that? If not, could it?

plabadens commented:
Unfortunately, the conversion script seems to break when applied to models generated using the latest version of https://github.com/qwopqwop200/GPTQ-for-LLaMa/commit/4c15f16aa7b9513983d03109da1b9479d7d0f11f. In this case, the `alpaca-native` model, quantized to 4bit with `--act-order` and `--groupsize 128`.
<details>
<summary>Output</summary>
<pre>
Loaded 'transformers' model split into 1 parts.
Writing vocab...
[1/291] Writing tensor tok_embeddings.weight, size 32001 x 4096...
[2/291] Writing tensor norm.weight, size 4096...
[3/291] Writing tensor output.weight, size 32001 x 4096...
Traceback (most recent call last):
  File "/home/pierre/Development/llama/llama.cpp/convert.py", line 673, in <module>
    main()
  File "/home/pierre/Development/llama/llama.cpp/convert.py", line 671, in main
    OutputFile.write_all(outfile, params, model, vocab)
  File "/home/pierre/Development/llama/llama.cpp/convert.py", line 579, in write_all
    for i, ((name, lazy_tensor), ndarray) in enumerate(zip(model.items(), ndarrays)):
  File "/home/pierre/Development/llama/llama.cpp/convert.py", line 508, in bounded_parallel_map
    result = futures.pop(0).result()
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/concurrent/futures/_base.py", line 451, in result
    return self.__get_result()
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/pierre/Development/llama/llama.cpp/convert.py", line 577, in <lambda>
    ndarrays = bounded_parallel_map(lambda lazy_tensor: lazy_tensor.load().ggml_ndarray(), model.values(),
  File "/home/pierre/Development/llama/llama.cpp/convert.py", line 357, in load
    tensor = lazy_tensor.load()
  File "/home/pierre/Development/llama/llama.cpp/convert.py", line 399, in load
    return QuantizedTensor(model, namebase)
  File "/home/pierre/Development/llama/llama.cpp/convert.py", line 187, in __init__
    scales = load_unquantized(model[f"{namebase}.scales"], np.float32)
  File "/home/pierre/Development/llama/llama.cpp/convert.py", line 178, in load_unquantized
    tensor = lazy_tensor.load()
  File "/home/pierre/Development/llama/llama.cpp/convert.py", line 473, in load
    return UnquantizedTensor(storage.load(storage_offset, elm_count).reshape(size))
ValueError: cannot reshape array of size 1 into shape (1,4096)
</pre>
</details>

comex commented:
> The memory layout on disk, should be the same as what ggml wants in memory at runtime. The format should also observe ideally a 32-byte alignment. Does this change do that? If not, could it?

The memory layout matches, but there is currently no alignment.  I was thinking of adding that, but it will require a format change, whereas this PR happens to be compatible with the existing format (since there was already an option to adjust the per-file split), so I decided to leave it out of this one. 

Regarding other feedback, I’ll take a look soon.

I’m also thinking about adding support for reading files that are already in GGML format so that they can be upgraded without needing the original.  This is despite the fact that I think it’s probably advisable to make the loader backwards-compatible moving forward rather than requiring upgrades.  Even with a change to add mmap support, there should be a fallback path that supports existing non-aligned files.  But if you want to actually *benefit* from mmap, you’ll need alignment and thus a format upgrade.

slaren commented:
Would it be possible to ensure that the tensor data is aligned by padding the tensor names with zeros? That should us to do it without changing the file format.

comex commented:
I thought of that, but it seemed like an ugly hack for not much benefit.  It’s not *hard* to change the C++ side; it just seemed convenient to make it a separate change to avoid merge conflicts and such.

slaren commented:
Something that may also help (as suggested by xloem in discord) would be making sure that the tensors in the model file are in the same order as they are accessed during inference. This should especially help in systems without enough memory to keep the entire model in memory. I think it is already very close to being that way, but may be worth double checking.

jart commented:
@Belluxx Kind of yes. @slaren and I have been collaborating on [Redbean's](https://redbean.dev) Discord server, which has an #AI channel. There's no official chatroom for the llama.cpp project yet, however you're all welcome to join us on the Redbean Discord until that happens! https://discord.gg/AqSvHf4u

linouxis9 commented:
Does the new conversion script works better with generic pytorch models? (Such as https://huggingface.co/THUDM/chatglm-6b)
Thanks :-)

BadisG commented:
I can confirm it doesn't work with the new implementations of the GPTQ quantization.
I tried it with the gpt-x-alpaca-13b-native-4bit-128g.pt model which was converted this way
```
CUDA_VISIBLE_DEVICES=0 python llama.py ./models/chavinlo-gpt4-x-alpaca --wbits 4 --true-sequential --act-order --groupsize 128 --save gpt-x-alpaca-13b-native-4bit-128g.pt
```

I got this error:
```
D:\Large Language Models\CONVERTISSEURS\gptq to ggml>python GPTQ-to-GGML.py gpt-x-alpaca-13b-native-
4bit-128g.pt --vocab-dir TokenDIR
Loaded 'transformers' model split into 1 parts.
Writing vocab...
[1/363] Writing tensor tok_embeddings.weight, size 32001 x 5120...
[2/363] Writing tensor norm.weight, size 5120...
[3/363] Writing tensor output.weight, size 32001 x 5120...
Traceback (most recent call last):
  File "D:\Large Language Models\CONVERTISSEURS\gptq to ggml\GPTQ-to-GGML.py", line 673, in <module>

    main()
  File "D:\Large Language Models\CONVERTISSEURS\gptq to ggml\GPTQ-to-GGML.py", line 671, in main
    OutputFile.write_all(outfile, params, model, vocab)
  File "D:\Large Language Models\CONVERTISSEURS\gptq to ggml\GPTQ-to-GGML.py", line 579, in write_al
l
    for i, ((name, lazy_tensor), ndarray) in enumerate(zip(model.items(), ndarrays)):
  File "D:\Large Language Models\CONVERTISSEURS\gptq to ggml\GPTQ-to-GGML.py", line 508, in bounded_
parallel_map
    result = futures.pop(0).result()
  File "C:\Users\Utilisateur\AppData\Local\Programs\Python\Python310\lib\concurrent\futures\_base.py
", line 451, in result
    return self.__get_result()
  File "C:\Users\Utilisateur\AppData\Local\Programs\Python\Python310\lib\concurrent\futures\_base.py
", line 403, in __get_result
    raise self._exception
  File "C:\Users\Utilisateur\AppData\Local\Programs\Python\Python310\lib\concurrent\futures\thread.p
y", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "D:\Large Language Models\CONVERTISSEURS\gptq to ggml\GPTQ-to-GGML.py", line 577, in <lambda>

    ndarrays = bounded_parallel_map(lambda lazy_tensor: lazy_tensor.load().ggml_ndarray(), model.val
ues(),
  File "D:\Large Language Models\CONVERTISSEURS\gptq to ggml\GPTQ-to-GGML.py", line 357, in load
    tensor = lazy_tensor.load()
  File "D:\Large Language Models\CONVERTISSEURS\gptq to ggml\GPTQ-to-GGML.py", line 399, in load
    return QuantizedTensor(model, namebase)
  File "D:\Large Language Models\CONVERTISSEURS\gptq to ggml\GPTQ-to-GGML.py", line 187, in __init__

    scales = load_unquantized(model[f"{namebase}.scales"], np.float32)
  File "D:\Large Language Models\CONVERTISSEURS\gptq to ggml\GPTQ-to-GGML.py", line 181, in load_unq
uantized
    assert tensor.ndarray.dtype == expected_dtype, (tensor.ndarray.dtype, expected_dtype)
AssertionError: (dtype('float16'), <class 'numpy.float32'>)
```

comex commented:
Updated:
- Now compatible with the merged mmap changes.
- Supports more formats (letting it serve as a replacement to all the existing scripts)...
  - although the latest GPTQ models are not supported yet, see PR message.
- Improved internals and added debug options.
- Added stub for convert-pth-to-ggml.py (since it's much older than the other scripts).

comex commented:
> Does the new conversion script works better with generic pytorch models? (Such as https://huggingface.co/THUDM/chatglm-6b)
> Thanks :-)

No, this codebase is still quite specific to LLaMA and derivatives.

BadisG commented:
@comex looks like there's 2 branches on the GPTQ-for-Llama repository

https://github.com/qwopqwop200/GPTQ-for-LLaMa/tree/triton
https://github.com/qwopqwop200/GPTQ-for-LLaMa/tree/cuda

And both propose the same quantization implementations, true sequential, act or order and group size

Do you think your script will work with both of them or you'll have to write a different code for each one of them? 😅

comex commented:
I *think* they use the same format.

qwopqwop200 commented:
> I _think_ they use the same format.

They use the same format.

comex commented:
Update: Tested with all three variants of `alpaca-30b-lora-int4`.  They all seem to work, including the 128g GPTQ version, because it was made using a slightly older version of GPTQ-for-LLaMa (it has the quantized zeros but not `g_idx`).

BadisG commented:
@qwopqwop200 Can you tell me what's the difference between the triton and cuda quantization? Which one should I choose at the end?

qwopqwop200 commented:
> @qwopqwop200 Can you tell me what's the difference between the triton and cuda quantization? Which one should I choose at the end?

Both are equally quantized. The only difference is whether to use the CUDA kernel or triton when running the model.

luxtiasco commented:
Any hint on how to make gptq 4bit work currently? Latest without --act-order, --true-sequential or --groupsize, seems not working. 
Or is it possible with an old version of gptq-llama?

xor2003 commented:
Does it supports adapter_model.bin?

TheBloke commented:
Thank you, this script is great and much improved over the previous conversion scripts. I used it today to convert Koala 7B (unquantized) from HF format to GGML. The older convert script failed because the data was in two .bin files and I didn't have a params.json that matched that (Koala 7B is 13GB in two files vs the original Llama 7B at 9GB in one file).  The new script worked perfectly first time, and it's awesome it does the PR613 conversion at the same time.

As many others have mentioned, it'd be awesome if this could also support the new GPTQ code. I was just about to run the Koala 7B and 13B models through GPTQ, until I realised that I wouldn't then be able to convert them to GGML. Instead I will try an older version of GPTQ in the hope that will work.

Thanks again for all the work you guys are doing on home LLMs!

TheBloke commented:
OK I will be experimenting with that soon and will update if I can figure it out.

chakflying commented on convert.py:
```diff
@@ -0,0 +1,1053 @@
+from typing import Optional, Callable, Type, Any, Iterable, IO, Sequence, Union, TypeVar, Literal
+from abc import abstractmethod, ABCMeta
+
+faulthandler.register(signal.SIGUSR1)
```
`faulthandler.register` does not exist on Windows, will throw AttributeError. ([link to doc](https://docs.python.org/3/library/faulthandler.html#dumping-the-traceback-on-a-user-signal))

Maybe we can reference this [PR](https://github.com/ipython/ipykernel/pull/108) in ipykernel for how to handle this, or just skip it?

TheBloke commented:
> OK I will be experimenting with that soon and will update if I can figure it out.

Update:
1. I looked through the code at https://github.com/qwopqwop200/GPTQ-for-LLaMa in the `cuda` branch, to find the latest commit that did not use `g_idx`
2. The commit I identified was [0a3fc17](https://github.com/qwopqwop200/GPTQ-for-LLaMa/commit/0a3fc179081e383054a555045f89c5b57ae2a6e9).  The next day they pushed the update that first added g_idx.
3. I compiled this code and ran it in a Google Colab. It failed due to error `'NoneType' object is not subscriptable`
4. After a bit of Googling, I found this Issue: https://github.com/huggingface/transformers/issues/22407 which resulted in this PR in `transformers`: https://github.com/huggingface/transformers/pull/22409 . That PR was closed without action because the GPTQ-for-Llama guys fixed it on their end instead, but that was in a later commit than I could use.  I wanted to keep things simple so I applied the transformers fix from their PR to my local transformers
5. With that done, I got a successful GPTQ 4bit 128g run of Koala 7B
6. This GPTQ appears to work, as confirmed by testing it with GPTQ-for-Llama's `llama_for_inference.py`:
![image](https://user-images.githubusercontent.com/784313/230499811-e71a97b3-be05-457b-8729-65884111792a.png)
7. However, converting it to GGML using the `convert.py` in this PR results in corrupted/garbage output from llama.cpp:
```
 $ ~/anaconda3/envs/torch21/bin/python ./convert.py  ~/src/llama.cpp/models/koala/7B/koala-7B-4bit-128g.olderFormat.pt --outtype q4_1 --outfile ~/src/llama.cpp/models/koala/7B/koala-7B-4bit-128g.ggml.bin
Loading model file /Users/tomj/src/llama.cpp/models/koala/7B/koala-7B-4bit-128g.olderFormat.pt
Loading vocab file /Users/tomj/src/llama.cpp/models/koala/7B/tokenizer.model
Writing vocab...
[1/291] Writing tensor tok_embeddings.weight, size 32000 x 4096...
[2/291] Writing tensor norm.weight, size 4096...
....
[291/291] Writing tensor layers.31.ffn_norm.weight, size 4096...
Wrote /Users/tomj/src/llama.cpp/models/koala/7B/koala-7B-4bit-128g.ggml.bin

$ ./main -t 18 -m models/koala/7B/koala-7B-4bit-128g.ggml.bin --color  -c 2048 --temp 0.7 --repeat_penalty 1.1 -n -1 -i -ins
main: seed = 1680816810
llama_model_load: loading model from 'models/koala/7B/koala-7B-4bit-128g.ggml.bin' - please wait ...
llama_model_load: n_vocab = 32000
llama_model_load: n_ctx   = 2048
llama_model_load: n_embd  = 4096
llama_model_load: n_mult  = 256
llama_model_load: n_head  = 32
llama_model_load: n_layer = 32
llama_model_load: n_rot   = 128
llama_model_load: f16     = 4
llama_model_load: n_ff    = 11008
llama_model_load: n_parts = 1
llama_model_load: type    = 1
llama_model_load: ggml map size = 5133.45 MB
llama_model_load: ggml ctx size =  81.25 KB
llama_model_load: mem required  = 6925.53 MB (+ 1026.00 MB per state)
llama_model_load: loading tensors from 'models/koala/7B/koala-7B-4bit-128g.ggml.bin'
llama_model_load: model size =  5133.02 MB / num tensors = 291
llama_init_from_file: kv self size  = 1024.00 MB

system_info: n_threads = 18 / 36 | AVX = 1 | AVX2 = 1 | AVX512 = 1 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 |
main: interactive mode on.
Reverse prompt: '### Instruction:

'
sampling: temp = 0.700000, top_k = 40, top_p = 0.950000, repeat_last_n = 64, repeat_penalty = 1.100000
generate: n_ctx = 2048, n_batch = 8, n_predict = -1, n_keep = 2


== Running in interactive mode. ==
 - Press Ctrl+C to interject at any time.
 - Press Return to return control to LLaMa.
 - If you want to submit another line, end your input in '\'.


> Write a story about Kevin
źattanörtérez corso Alb Tob estavenurdalla fitérezyerunity dressurd Savpois fit oscidi� SEhttpsalla Girlsẓlarunnelalla corso fiturdollyunityurd�érez Podpoisussenyenunn Savurd girlspaunity owRefRef танur�éliunitymockallagun�urdurd finalefitussenźurdidiattanattanRefpois Sieg Normanurdurd fit corsourd Alb Siegört dressurdyer Tob osc dressurd�ussen viörtRefpoisörtunityunneltx SE Intpoislarattan fitsmock�debugattanérez Albpa Savurd girls^C
```

One thing I noticed is that the GGML produced by `convert.py` was 30% larger than the input pt file - is this expected?
```
-rw-r--r--   1 tomj  staff   5.0G  6 Apr 22:32 koala-7B-4bit-128g.ggml.bin
-rw-------@  1 tomj  staff   3.5G  6 Apr 21:52 koala-7B-4bit-128g.olderFormat.pt
```
So I'm not sure what to do now.  Could this be a bug in the `convert.py` script? Or have I done something wrong?

If anyone (@comex ?) would be willing to look at this I can make the source files accessible for download.

I'm still a newbie when it comes to AI and ML file formats so I'm not really sure how to debug further myself..

Belluxx commented:
@TheBloke Yes the larger file size is expected behaviour, i got the same resulting size after conversion, but my model worked fine and was coherent.

TheBloke commented:
> @TheBloke Yes the larger file size is expected behaviour, i got the same resulting size after conversion, but my model worked fine and was coherent.

OK thanks for confirming. Have you tried running convert.py on any GPTQ files at all?

If I can't get GPTQ to work then the alternative would be to use the `./quantize` that comes with llama.cpp, which I guess is the older algorithm that isn't as good. But at least it works!

TheBloke commented:
Today I was trying to run the new convert.py from this PR on a Google Colab instance, which uses Python 3.9.16.

It's failing like so:

```
/content# python convert.py /content/koala-7B-HF --outfile /content/ggml.bin
Loading model file /content/koala-7B-HF/pytorch_model-00001-of-00002.bin
Traceback (most recent call last):
  File "/content/convert.py", line 1053, in <module>
    main()
  File "/content/convert.py", line 1034, in main
    model_plus = load_some_model(args.model)
  File "/content/convert.py", line 966, in load_some_model
    models_plus.append(lazy_load_file(path))
  File "/content/convert.py", line 771, in lazy_load_file
    return lazy_load_torch_file(fp, path)
  File "/content/convert.py", line 653, in lazy_load_torch_file
    model = unpickler.load()
TypeError: 'staticmethod' object is not callable

/content# python -V
Python 3.9.16
```

Running the same command, with the same model data, runs fine on my local macOS system using both Python 3.10.10 (Anaconda install) and 3.11.2 (main Homebrew install.)

I'm not very experienced with Python so I can't figure out if this is an environment problem on the Colab, or perhaps a bug in the code specific to running on an earlier version of Python?

Thanks in advance for any help/advice.

ggerganov reviewed (approved):
Looks like this is ready to merge.
If so, @comex please go ahead

DannyDaemonic commented:
I tested this patch on my older Ubuntu system with Focal Fossa (20.04 LTS) installed, but it didn't work due to a compatibility issue. The code uses a feature introduced in Python 3.9, specifically type hinting generics in standard collections:
```python
NDArray = np.ndarray[Any, Any]
```
This was not compatible with Python 3.8 installed on my old system. I updated my Ubuntu to Jammy Jellyfish (22.04 LTS), which installed Python 3.10 and resolved the compatibility issue.

However, I initially used the older system because my newer systems have Python 3.11 installed. Unfortunately, sentencepiece still has wheel problems with Python 3.11, so the code doesn't run on those either.

This new code is walking a fine line by requiring a Python version newer than 3.8 but earlier than 3.11. It's probably not a big deal, but it might be something to consider at some point to ensure the code works across a broader range of systems.

TheBloke commented:
> This new code is walking a fine line by requiring a Python version newer than 3.8 but earlier than 3.11. It's probably not a big deal, but it might be something to consider at some point to ensure the code works across a broader range of systems.

Yeah I agree.  And there appear to be issues on Python 3.9 as well - I posted earlier about how the script fails on Google Colab, running Python 3.9.16.  

I think I have confirmed that this was a version-specific issue by upgrading the Colab to Python 3.10.6, which enables the script to run fine.  I'm glad I can get the script working in Colab, but it does require that I first prepare each Colab instance by running a Python upgrade, which adds extra setup time.

It would be really great if this script could be a little more flexible in the Python versions it would run with.  

I'm also thinking from a perspective of helping users, eg on Discord. I can see people coming to llama.cpp from all sorts of systems, perhaps not updated for a little while, and it'll complicate things if the script won't work except on a narrow range of Python versions.

It's an awesome script which is so much better than the old ones, so it'd be awesome if as many people could run it as possible, as easily as possible.

TheBloke commented:
> I can confirm the script doesn't work on Google Colab (Python version 3.9.16).
> 
> > I think I have confirmed that this was a version-specific issue by upgrading the Colab to Python 3.10.6.
> 
> How did you do that? I found a way on SO, but it involves running a 348MB `*.sh` file I'm not comfortable running.

That's exactly how I did it :)
```
!wget https://github.com/korakot/kora/releases/download/v0.10/py310.sh
!bash ./py310.sh -b -f -p /usr/local
!python -m ipykernel install --name "py310" --user
```
It's a temporary instance in the cloud and I'm not putting any personal data into it, so I didn't have any great worries. Google Colab prompts you for permission if any code tries to mount your Google Drive, so I figured there wasn't a huge risk. Also it's been around a while and seems to be highly regarded as a repo.

Mihaiii commented:
@TheBloke  Thanks, I also tried that eventually!

I bypassed the initial model loading error, but unfortunately I still can't convert using the script.
Not clear if it's a script problem or something I do wrong.

```
Traceback (most recent call last):
  File "/content/gdrive/MyDrive/LearningGraph/llama-cpp-python/vendor/llama.cpp/convert.py", line 1053, in <module>
    main()
  File "/content/gdrive/MyDrive/LearningGraph/llama-cpp-python/vendor/llama.cpp/convert.py", line 1034, in main
    model_plus = load_some_model(args.model)
  File "/content/gdrive/MyDrive/LearningGraph/llama-cpp-python/vendor/llama.cpp/convert.py", line 968, in load_some_model
    model_plus = merge_multifile_models(models_plus)
  File "/content/gdrive/MyDrive/LearningGraph/llama-cpp-python/vendor/llama.cpp/convert.py", line 519, in merge_multifile_models
    model = merge_sharded([mp.model for mp in models_plus])
  File "/content/gdrive/MyDrive/LearningGraph/llama-cpp-python/vendor/llama.cpp/convert.py", line 499, in merge_sharded
    return {name: convert(name) for name in names}
  File "/content/gdrive/MyDrive/LearningGraph/llama-cpp-python/vendor/llama.cpp/convert.py", line 499, in <dictcomp>
    return {name: convert(name) for name in names}
  File "/content/gdrive/MyDrive/LearningGraph/llama-cpp-python/vendor/llama.cpp/convert.py", line 475, in convert
    lazy_tensors: list[LazyTensor] = [model[name] for model in models]
  File "/content/gdrive/MyDrive/LearningGraph/llama-cpp-python/vendor/llama.cpp/convert.py", line 475, in <listcomp>
    lazy_tensors: list[LazyTensor] = [model[name] for model in models]
KeyError: 'transformer.word_embeddings.weight'
```

qwopqwop200 commented:
act-order doesn't work, but I changed the code to make the latest GPTQ work.
(https://github.com/qwopqwop200/llama.cpp/blob/master/convert.py)
Currently, I believe the following improvements are possible:
1. If QK set to 128, it can reduce 6 bit to 4.5bit
2. if it change to support act-order. it get better performance

CRD716 commented:
```
Traceback (most recent call last):
  File "E:\largefiles\LLaMA\convert.py", line 10, in <module>
    faulthandler.register(signal.SIGUSR1)
AttributeError: module 'faulthandler' has no attribute 'register'
```
Failing to run on Windows 10

comex commented:
Sorry for not attending to this.  Let me deal with the feedback that's accumulated…

qwopqwop200 commented:
> act-order doesn't work, but I changed the code to make the latest GPTQ work. (https://github.com/qwopqwop200/llama.cpp/blob/master/convert.py) Currently, I believe the following improvements are possible:
> 
> 1. If QK set to 128, it can reduce 6 bit to 4.5bit
> 2. if it change to support act-order. it get better performance

The first one confirmed that doesn't work.

comex committed:
New conversion script (#545)

  Current status: Working, except for the latest GPTQ-for-LLaMa format
  that includes `g_idx`.  This turns out to require changes to GGML, so
  for now it only works if you use the `--outtype` option to dequantize it
  back to f16 (which is pointless except for debugging).

  I also included some cleanup for the C++ code.

  This script is meant to replace all the existing conversion scripts
  (including the ones that convert from older GGML formats), while also
  adding support for some new formats.  Specifically, I've tested with:

  - [x] `LLaMA` (original)
  - [x] `llama-65b-4bit`
  - [x] `alpaca-native`
  - [x] `alpaca-native-4bit`
  - [x] LLaMA converted to 'transformers' format using
        `convert_llama_weights_to_hf.py`
  - [x] `alpaca-native` quantized with `--true-sequential --act-order
        --groupsize 128` (dequantized only)
  - [x] same as above plus `--save_safetensors`
  - [x] GPT4All
  - [x] stock unversioned ggml
  - [x] ggmh

  There's enough overlap in the logic needed to handle these different
  cases that it seemed best to move to a single script.

  I haven't tried this with Alpaca-LoRA because I don't know where to find
  it.

  Useful features:

  - Uses multiple threads for a speedup in some cases (though the Python
    GIL limits the gain, and sometimes it's disk-bound anyway).

  - Combines split models into a single file (both the intra-tensor split
    of the original and the inter-tensor split of 'transformers' format
    files).  Single files are more convenient to work with and more
    friendly to future changes to use memory mapping on the C++ side.  To
    accomplish this without increasing memory requirements, it has some
    custom loading code which avoids loading whole input files into memory
    at once.

  - Because of the custom loading code, it no longer depends in PyTorch,
    which might make installing dependencies slightly easier or faster...
    although it still depends on NumPy and sentencepiece, so I don't know
    if there's any meaningful difference.  In any case, I also added a
    requirements.txt file to lock the dependency versions in case of any
    future breaking changes.

  - Type annotations checked with mypy.

  - Some attempts to be extra user-friendly:

      - The script tries to be forgiving with arguments, e.g. you can
        specify either the model file itself or the directory containing
        it.

      - The script doesn't depend on config.json / params.json, just in
        case the user downloaded files individually and doesn't have those
        handy.  But you still need tokenizer.model and, for Alpaca,
        added_tokens.json.

      - The script tries to give a helpful error message if
        added_tokens.json is missing.

comex commented:
Updates:
- Fixed Python 3.8 compatibility.  (By the way, installing sentencepiece on Python 3.11 also works fine for me, but maybe it depends on the OS.)
- Fixed `faulthandler` incompatibility with Windows.
- Fixed `TypeError: 'staticmethod' object is not callable`.
- Fixed error if `scales` is fp16 instead of fp32 (Koala in dequantize mode).

This weekend hopefully I'll get to fixing compatibility with the latest GPTQ.

comex commented:
Looks like GitHub doesn't give me a merge button even with the approval and checks passing (not sure why), but feel free to merge, @ggerganov.  Thanks!

ggerganov commented:
@comex 

Thank you for the hard work and for another very well done contribution!

DannyDaemonic commented:
Since `requirements.txt` is going into the root directory, to avoid confusion we should consider renaming it to something like `convert-reqs.txt` or `conversion-requirements.txt`, as those requirements are specific to the conversion scripts.

vmajor commented:
How can convert.py be used to migrate old ggml model to the new ggml model? Attempting to do so blindly results in this error:

`python convert.py  --outfile ../alpaca.cpp_65b_ggml/new_ggml-model-q4_0.bin ../alpaca.cpp_65b_ggml/ggml-model-q4_0.bin`

```
raise FileNotFoundError(f"Could not find tokenizer.model in {path} or its parent; if it's in another directory, pass the directory as --vocab-dir")
FileNotFoundError: Could not find tokenizer.model in ../alpaca.cpp_65b_ggml or its parent; if it's in another directory, pass the directory as --vocab-dir
```

TheBloke commented:
> How can convert.py be used to migrate old ggml model to the new ggml model? Attempting to do so blindly results in this error:
> 
> `python convert.py --outfile ../alpaca.cpp_65b_ggml/new_ggml-model-q4_0.bin ../alpaca.cpp_65b_ggml/ggml-model-q4_0.bin`
> 
> ```
> raise FileNotFoundError(f"Could not find tokenizer.model in {path} or its parent; if it's in another directory, pass the directory as --vocab-dir")
> FileNotFoundError: Could not find tokenizer.model in ../alpaca.cpp_65b_ggml or its parent; if it's in another directory, pass the directory as --vocab-dir
> ```

You can download the tokenizer.model it's missing from HF, eg at this link: https://huggingface.co/TheBloke/alpaca-lora-65B-HF/resolve/main/tokenizer.model

PS. If you want a newer 65B Alpaca Lora model, using newer and better 4bit quantisation techniques, try the q4_0, q4_2 or q4_3 models from my repo here: https://huggingface.co/TheBloke/alpaca-lora-65B-GGML .  q4_2 seems to be the quantisation format that people regard as best at the moment.

vmajor commented:
Thank you for this! I still cannot get the conversion done due to a different error, but I downloaded your model and it is now working much better. Can you tell me what is the difference from the later q4_3 model? That one is larger.

Green-Sky commented:
@vmajor https://github.com/ggerganov/llama.cpp/discussions/1121

big-thousand commented:
which version of GPTQ-for-LLaMa can get no g_idx model.

Green-Sky commented:
@big-thousand I believe you are in the wrong repository.

Interpause commented:
Would like to check if there is now support for converting GPTQ 4-bit quantized models to GGML

TheBloke commented:
I think they're asking because llama.cpp convert.py can convert old GPTQ models to GGML, but only if they don't have the new GPTQ `g_idx` format.

However @big-thousand and @Interpause I do not recommend you do this.  I tested using convert.py to convert GPTQ -> GGML and the perplexity (model accuracy) was very poor.  Much worse than using llama.cpp's own quantize feature.

I think this is partly because you have to use an old version of GPTQ to do the conversion.

I suggest you make new GGMLs using llama.cpp `quantize`. It will result in the highest quality model, and will be faster than going float16 -> GPTQ -> GGML.

That's what I do now for all my model releases on HF.  I do float16 -> GPTQ, and separately I do float16 -> GGML

Interpause commented:
Just would like to ask, current GGML 4 bit does some form of error correction right? The main rationale behind wanting to use GPTQ is to mitigate increase in perplexity. Is GGML's 4 bit quantization already on par or superior to GPTQ?

TheBloke commented:
> Just would like to ask, current GGML 4 bit does some form of error correction right? The main rationale behind wanting to use GPTQ is to mitigate increase in perplexity. Is GGML's 4 bit quantization already on par or superior to GPTQ?

I am actually testing that right this second.  I wrote a perplexity calc for GPTQ that runs 100% the same algorithm as the `perplexity` tool in llama.cpp, so the results are comparable.

Here are some early results from my testing (which I will publish properly soon):

Llama 7B:
* float16: 5.9066
* llama.cpp q4_0: 6.1565
* llama.cpp q4_1: 6.0910
* llama.cpp q5_0: 5.9862
* llama.cpp q5_1: 5.9481
* llama.cpp q8_0: 5.9069
* AutoGPTQ 4bit 32g no desc_act: 6.2650
* AutoGPTQ 4bit 32g desc_act: 6.0422
* AutoGPTQ 4bit 128g no desc_act: 6.3850
* AutoGPTQ 4bit 128g desc_act: 6.0653
 
So you see that for 4bit, GPTQ is slightly better. Best result is 6.0422 or 6.0653.  Although this requires `desc_act`, and there are currently some performance implications to using that - it slows down inference a fair bit at the moment.

But llama.cpp also offers 5bit, and this out-performs GPTQ 4bit. And now that llama.cpp has CUDA GPU acceleration, it may be it can compete on performance as well.

So it will be up to the user to decide what is best for them and their use case. 

I will publish more results, and benchmarks, soon.

ggerganov commented:
@TheBloke - When doing the comparisons, don't forget to include the file sizes. These are important

YellowRoseCx commented:
I get this error with the new script, trying to convert FP16 models

```
$ python convert.py "/mnt/UbuntuData/AI/Models/pygmalion-13b/pygmalion-13b" --outtype f16
Loading model file /mnt/UbuntuData/AI/Models/pygmalion-13b/pygmalion-13b/pytorch_model-00001-of-00003.bin
Loading model file /mnt/UbuntuData/AI/Models/pygmalion-13b/pygmalion-13b/pytorch_model-00001-of-00003.bin
Loading model file /mnt/UbuntuData/AI/Models/pygmalion-13b/pygmalion-13b/pytorch_model-00002-of-00003.bin
Loading model file /mnt/UbuntuData/AI/Models/pygmalion-13b/pygmalion-13b/pytorch_model-00003-of-00003.bin
Traceback (most recent call last):
  File "/mnt/UbuntuData/AI/koboldcpp/convert.py", line 1168, in <module>
    main()
  File "/mnt/UbuntuData/AI/koboldcpp/convert.py", line 1148, in main
    model_plus = load_some_model(args.model)
  File "/mnt/UbuntuData/AI/koboldcpp/convert.py", line 1074, in load_some_model
    models_plus.append(lazy_load_file(path))
  File "/mnt/UbuntuData/AI/koboldcpp/convert.py", line 864, in lazy_load_file
    return lazy_load_torch_file(fp, path)
  File "/mnt/UbuntuData/AI/koboldcpp/convert.py", line 729, in lazy_load_torch_file
    zf = zipfile.ZipFile(outer_fp)
  File "/usr/lib/python3.10/zipfile.py", line 1267, in __init__
    self._RealGetContents()
  File "/usr/lib/python3.10/zipfile.py", line 1334, in _RealGetContents
    raise BadZipFile("File is not a zip file")
zipfile.BadZipFile: File is not a zip file

```

earonesty commented:
this should be its own release on pypi
